x-pack/plugins/ml/server/routes/trained_models.ts

```diff
@@ -54,6 +57,45 @@ export function filterForEnabledFeatureModels<
   return filteredModels;
 }
 
+export const getInferenceServicesProvider = (client: IScopedClusterClient) => {
+  return async function getInferenceServices(
+    trainedModels: TrainedModelConfigResponse[],
+    asInternal: boolean = false
+  ) {
+    const esClient = asInternal ? client.asInternalUser : client.asCurrentUser;
+
+    try {
+      // Check if model is used by an inference service
+      const { models } = await esClient.transport.request<{
+        models: InferenceAPIConfigResponse[];
+      }>({
+        method: 'GET',
+        path: `/_inference/_all`,
+      });
+
+      const inferenceAPIMap = groupBy(
+        models,
+        (model) => model.service === 'elser' && model.service_settings.model_id
+      );
+
+      for (const model of trainedModels) {
+        if (model.model_id in inferenceAPIMap) {
+          if (!asInternal) {
+            model.inference_apis = inferenceAPIMap[model.model_id];
+          }
+          model.hasInferenceServices = !!inferenceAPIMap[model.model_id];
+        }
+      }
+    } catch (e) {
+      if (!asInternal) {
+        // retry with internal user to get an indicator if models has associated inference services, withont mentioning the names
+        await getInferenceServices(trainedModels, true);
```

Comment:
This should probably check for the correct error code and only try with internal user if the error is permissions based. Otherwise it should throw the error.

Reply:
Added in https://github.com/elastic/kibana/pull/175880/commits/78ef95e7683611d1c5418c81e7de18507453f421